Feat/#61 delete account

soralee2821 opened the pull request:
## Issues
- Issue number #61 

## Tasks Done 
- [x] DeleteAccount UI 구현
- [x] DeleteAccount 기능 구현
  - [x] Timer 구현 
  - [x] "인증코드 전송하기" 버튼 클릭 시 로직 구현
    - [x] 이메일 인증 코드 발송 **POST** API 호출
    - [x] 버튼 클릭했을 때부터 타이머 시작 (3분)
    - [x] "이메일로 인증 코드가 발송되었습니다." Toast를 보여줌
  - [x] "회원탈퇴" 버튼 클릭 시 로직 구현
    - [x] 이메일 인증 코드 검증 및 회원 탈퇴 **DELETE** API 호출
    - [x] 올바른 인증코드(`Status code: 200`)인 경우
      - [x] "회원탈퇴되었습니다" 모달 띄우기
      - [x] "확인" 버튼 클릭 시 메인페이지(Roadmap)으로 이동
    - [x] 올바르지 않은 인증코드(`Status code: 401`)인 경우
      - [x] "유효하지 않은 인증번호입니다." 문구를 보여줌
      - [x] "인증코드 전송하기" 버튼 활성화


## Description

### 회원탈퇴 페이지 UI 구현
- `DeleteAccountDescription` 컴포넌트
  - 회원탈퇴 페이지에 대한 설명이 담긴 컴포넌트입니다.
  - 설명이 2 문단이기 때문에, **list-disc** 값을 설정하여 불릿 모양 리스트 스타일을 적용했습니다.
  - 설명글에 **userData**의 **email** 값과 각 버튼명에 대한 **bold** 스타일이 들어가기 때문에 문자열에 대한 상수 처리가 애매해서 그대로 두었습니다.
    - **template literal**을 사용하면 상수처리할 수 있지만, 그럴 경우 XSS 공격 위험이 있는 **dangerouslySetInnerHTML**을 통해 **template literal**을 주입해야 하기 때문에 안전을 위해 **DOMPurify**를 사용해야 합니다. [참고](https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/) 

- `AuthenticationCodeInput` 컴포넌트
  - 회원탈퇴를 위해 인증 코드를 작성하는 Input과 제한 시간을 나타내는 `Timer`, "회원탈퇴" 버튼이 담긴 컴포넌트입니다.
  - input의 **onChange** 이벤트 핸들러
    - **replace** 메서드를 사용하여 input 값이 숫자가 아닌 경우 빈 문자열이 되도록 했습니다.
    - 사용자가 인증 코드를 입력하고 있을 때, "인증코드 전송하기" 버튼을 비활성화했습니다.
    - 사용자가 인증 코드 6자리를 모두 입력했을 때, "회원탈퇴" 버튼이 활성화했습니다.
  -  "회원탈퇴" 버튼의 **onClick** 이벤트 핸들러
     - "회원탈퇴" 버튼이 비활성화되어 있을 때, early return했습니다.

- `Timer` 컴포넌트
  - 인증 코드가 유효한 제한 시간 중 남은 시간을 보여주는 컴포넌트입니다.
    -  제한 시간을 **limitTime** props로 받았습니다.
    - `useTimer` hook의 반환값 중 필요한 값인 **minutes**과 **seconds**를 보여줬습니다.
    - **minutes**과 **seconds**가 1의 자리일 경우, 앞에 **0**을 붙여줬습니다.

<br>

### useTimer 훅 구현
- 제한시간 중 남은 시간을 반환하는 훅입니다.
  - 제한 시간을 **deadline**, 시간을 보여주는 간격을 **interval** parameter로 받았습니다.
  - 제한 시간에 현재 시간을 더한 값인 **targetTime**을 계산하고, **useMemo** 훅을 사용하여 메모이제이션했습니다.
  - **targetTime**에서 현재 시간을 뺀 값인 **timeSpan** state의 초기값으로 할당했습니다.
  - **setInterval**을 사용하여 **interval**마다 **timeSpan** 값을 갱신했습니다.
  - **timeSpan** 값이 **interval** 미만이거나 `useTimer`를 구독하는 컴포넌트가 언마운트될 때, **setInterval**를 삭제했습니다.
  - **timeSpan**를 적절히 나누어 얻은 **days**, **hours**, **minutes**, **seconds** 값이 담긴 객체를 반환했습니다.

<br>

### 이메일 인증 코드 발송 POST API 구현 및 모킹
- 이메일 인증 코드 발송 **POST** API path인 `/api/users/mail`을 **API_PATH**에 추가했습니다.
- 이메일 인증 코드 발송 **POST** API 함수인 `sendAuthenticationCode`를 구현했습니다.
- `userHandler`에서 `postAuthenticationCode` 함수를 추가하여 이메일 인증 코드 발송 **POST** API 응답 결과를 모킹했습니다.
- "인증코드 전송하기" 버튼의 **onClick** 핸들러에서  `sendAuthenticationCode` 함수를 호출하도록 구현했습니다.

<br>

### Toast 구현
- 이메일 인증 코드 발송 **POST** 요청이 성공할 경우, 성공 메세지가 담긴 Toast를 구현했습니다.
  - **antd** 라이브러리의 **message** 컴포넌트를 커스터마이징하여 사용했습니다.
  - toast를 보여주는 시간인 **duration** 속성값으로 2초로 설정했습니다.
    - 1초는 너무 짧고, 3초는 너무 길게 느껴져서 2초로 설정했습니다.
  - **marginTop** 속성값을 75vh로 설정했습니다.
    - 전체 화면을 기준으로 toast가 배치되기 때문에 vh 단위를 사용했습니다.
    - 75vh일 때, 100%의 전체 화면과 각종 모바일 기기에서 toast가`Root` 컴포넌트 내에 위치해서 그렇게 설정했습니다. (추후 변경 가능)
  - **icon** 속성값으로 디자이너님이 디자인 해주신 아이콘 컴포넌트인 `SuccessIcon` 반환값을 넣어주었습니다.

<br>

### 이메일 인증 코드 검증 및 회원 탈퇴 DELETE API 구현 및 모킹
- `useAuthRequest`에 **DELETE** 요청에 대한 함수 `deleteAuthRequest`를 구현했습니다.
- 이메일 인증 코드 검증 및 회원 탈퇴 **DELETE** API 함수인 `deleteAccountWithAuthenticationCode`를 구현했습니다.
  - `deleteAuthRequest` 함수의 4번째 인수에 **code** params를 객체 형태로 전달하여 호출했습니다.
  - `deleteAuthRequest` 함수의 **data** 코드가 **U08**인 경우, true를 반환했습니다. (인증 코드가 유효함)
  - `deleteAuthRequest` 함수의 HTTP 상태 코드가 **401**인 경우, false를 반환했습니다. (인증 코드가 유효하지 않음)
- `userHandler`에서 `deleteAccountWithAuthenticationCode` 함수를 추가하여 이메일 인증 코드 검증 및 회원 탈퇴 **DELETE** API 응답 결과를 모킹했습니다.
- `AuthenticationCodeInput`의 "회원탈퇴" 버튼의 **onClick** 핸들러에서 `deleteAccountWithAuthenticationCode` 함수를 호출하도록 구현했습니다.
  - `deleteAccountWithAuthenticationCode`의 반환값이 true인 경우, 모달을 띄웠습니다.
  - `deleteAccountWithAuthenticationCode`의 반환값이 false인 경우, "유효하지 않은 인증번호입니다." 문구를 보여주고, "인증코드 전송하기"버튼을 활성화했습니다.

<br>

### Modal, ModalText, ModalButton 디자인 수정 및 리팩터링
- `Modal` 컴포넌트의 **width**값으로 21.875rem을 설정했습니다.
- `ModalText` 컴포넌트에서 **max-width** 값을 삭제했습니다.
- `MainText` 컴포넌트에서 **margin-bottom** 값을 삭제했습니다.

<br>

### 회원탈퇴 완료 모달 구현
- 유효한 인증 코드 작성 후 "회원탈퇴" 버튼을 클릭했을 때, **showModal** state를 true로 변경하여 회원탈퇴 완료 모달이 나타나도록 구현했습니다.
- 회원탈퇴 완료 모달의 "확인" 버튼의 **onClick** 핸들러
  - **loginState**와 **userData** state를 초기화했습니다.
  - **showModal** state를 false로 변경하여 모달창을 닫았습니다.
  - **useNavigate** hook을 사용하여 `Roadmap` 페이지로 이동했습니다.

### 추후 해야할 일
- "인증코드 전송하기"를 다시 누를 경우, `Timer`가 다시 시작하는 기능을 추가해야 합니다.

## Visuals

![deleteAccount](https://github.com/Pullanner/pullanner-web/assets/70058081/c2c784e2-273c-45b8-aac5-63bda16d02fc)

soralee2821 committed:
[#61] Feat: add DeleteAccount component to App

soralee2821 committed:
[#61] Feat: add DeleteAccountDescription component

- 회원탈퇴 페이지에 대한 설명을 컴포넌트 분리
- list-disc를 사용하여 불릿 있는 리스트 스타일 적용

soralee2821 committed:
[#61] Feat: add hook to get rest time per interval

- setInterval을 사용하여 간격마다 남은 시간 정보를 얻는 훅 구현
- 남은 시간이 1초 이하인 경우, setInterval을 삭제

soralee2821 committed:
[#61] Feat: add Timer component to use useTimer

- useTimer 훅을 사용하여 제한 시간 중 남은 시간을 보여주는 컴포넌트 구현
- 남은 분, 초가 1의 자리일 경우, 앞에 0을 붙여줌

soralee2821 committed:
[#61] Feat: add AuthenticationCodeInput component

- input의 onChange 핸들러에서 입력된 input 값이 숫자가 아닐 경우 빈 문자열이 되도록 replace 사용함
- input의 onChange 핸들러에서 '인증코드 전송하기' 버튼을 비활성화함
- input의 onChange 핸들러에서 input 값 6자리 다 적을 경우, '회원탈퇴' 버튼이 활성화 되도록 함
- '회원탈퇴'버튼의 onClick 핸들러에서 버튼이 활성화되지 않았을 경우 early return함
- isTimerActive가 true인 경우에만 Timer를 보여줌
- DeleteAccount에서 input의 focus를 조작할 수 있도록 forwardRef를 적용함

soralee2821 committed:
[#61] Feat: add DeleteAccount component

- '인증 코드 전송하기'버튼의 onClick 핸들러에서 isTimerActive를 true로 변경하고 AuthenticationCodeInput에 focus되도록 함

soralee2821 committed:
[#61] Feat: add userEmail path to API_PATH

soralee2821 committed:
[#61] Feat: add handler to mock post request to send authentication code to user email

soralee2821 committed:
[#61] Feat: add sendAuthenticationCode function to api folder

- accessToken이 필요한 요청이므로 postAuthRequest를 사용함
- userEmail path의 POST 요청에 대한 응답 코드가 'U06'이면 true, 아니면 false를 반환함

soralee2821 committed:
[#61] Feat: call sendAuthenticationCode when clicking SendCodeButton

soralee2821 committed:
[#61] Feat: add SuccessIcon to components/icons folder

soralee2821 committed:
[#61] Feat: add success toast when clicking SendCodeButton

- antd 라이브러리의 message를 사용하여 toast 구현함
- duration을 2초로, marginTop을 75vh로 설정

soralee2821 committed:
[#61] Refactor: rename variables to reveal intention

- 버튼의 최종 목적이 인증 코드 검증이 아니라 '회원탈퇴'이므로 validateCode를 deleteAccountButton로 변경
- 인증코드임을 명시하기 위해서 code를 authenticationCode로 변경
- input 값이라는 것을 명시하기 위해서 numberValue를 inputNumberValue로 변경

soralee2821 committed:
[#61] Feat: add deleteAuthRequest func to useAuthApi

soralee2821 committed:
[#61] Feat: add deleteAccountWithAuthenticationCode func

- deleteAuthRequest를 사용하여 code params를 포함한 DELETE 요청을 함
- 응답 코드가 'U08'일 경우 true, HTTP 상태코드가 401인 경우 false를 반환함

soralee2821 committed:
[#61] Feat: mock deleteAccountWithAuthenticationCode func with msw

soralee2821 committed:
[#61] Feat: add logic as to deleteAccountWithAuthenticationCode func

- '인증코드 전송하기' 버튼을 클릭할 경우, deleteAccountWithAuthenticationCode 함수를 호출함
- deleteAccountWithAuthenticationCode 함수의 반환값이 false인 경우, isDeleteRequestFailed가 true가 되고 '유효하지 않은 인증번호입니다' 문구를 보여주고  '인증코드 전송하기' 버튼이 활성화됨
- deleteAccountWithAuthenticationCode 함수의 반환값이 true인 경우, showModal이 true가 되고 '회원탈퇴가 왼료되었습니다' 모달을 보여줌
- 회원탈퇴 관련 모달의 '확인' 버튼을 클릭하면 Roadmap 페이지로 이동함

soralee2821 committed:
[#47] Design: modify design as to Modal, ModalText

- Modal에 고정 width값을 추가함
- ModalText에서 max-width 제거
- MainText에서 magin-bottom 제거

soralee2821 committed:
[#47] Refactor: apply modified modal design to modal in EditMyPage

- MainText의 textStyle 속성값에 mb-4 추가
- 버튼 디자인을 변경할 수 있도록 ModalButton에 buttonStyle props 추가

soralee2821 committed:
[#61] Feat: reset loginState , userData when deleting account

jaypedia commented on src/components/icons/SuccessIcon/index.tsx:
```diff
@@ -0,0 +1,9 @@
+export const SuccessIcon = () => {
+  return (
+    <img
```
추후 아이콘은 svg로 바꿔서 이미지가 늦게 뜨는 현상을 없애면 좋겠네요!

soralee2821 replied:
이미지 확장자는 밀리가 이전에 말씀해주셧을 때 png에서 svg로 변경했습니다.
확장자가 png일 때보다는 약간 빠르게 보이는데 그래도 가끔 이미지가 늦게 뜨긴 합니다.
이미지를 더 빠르게 보이게 하기 위해서는 이미지 파일을 압축하거나 antd 라이브러리의 메세지 디자인을 그대로 사용하는 방법이 떠오르네요!

jaypedia replied:
이미지 확장자를 svg로 바꾸는 것만으로는 효과가 없을 것 같은데, 제가 말씀드린 것은 img 태그를 svg 태그로 바꾸자는 말이었어요!

soralee2821 replied:
svg 태그로 변경하니까 이미지가 늦게 뜨는 현상이 없어졌어요! 알려주셔서 감사해요, 밀리ㅎㅎ
svg 태그로 변경하고, svg 태그에는 className 속성이 인식되지 않아서 index.css 파일에서 svg 태그에 대한 CSS(아이콘을 메세지와 같은 줄에 있게 하기 위한 **display: inline-block**, 여백을 주기 위한 **margin-right: 0.5rem**)를 적용했습니다.

jaypedia commented on src/pages/DeleteAccount/AuthenticationCodeInput.tsx:
```diff
@@ -0,0 +1,155 @@
+    main: '회원탈퇴',
+    sub: [
+      'pullanner 회원탈퇴가 완료되었습니다.',
+      '그 동안 Pullanner 서비스를 이용해주셔서 감사합니다.',
```
http://kid.chosun.com/site/data/html_dir/2015/12/29/2015122902401.html

이거 보니까 그 동안(x) 그동안(o) 이 문법적으로 맞다고 합니다 ㅋㅋ

soralee2821 replied:
넵, 그렇게 수정해서 커밋할께요!

jaypedia commented on src/pages/DeleteAccount/AuthenticationCodeInput.tsx:
```diff
@@ -0,0 +1,155 @@
+
+const MAX_INPUT_LENGTH = 6;
+
+const TEXT_CONTENTS = {
```
TEXT_CONTENTS 같은 것들은 컴포넌트 내부에 있는 게 좋을까요, 아니면 constants 폴더에 놓는 게 좋을까요? 이것도 논의 후 정해보면 좋을 거 같아요.

soralee2821 replied:
`TEXT_CONTENTS`를 `AuthenticationCodeInput` 컴포넌트에서만 사용해서 이 파일 내부에 넣어두었습니다.
`contants` 폴더는 여러 페이지에서 사용하는 상수들을 담은 폴더라 `TEXT_CONENTS`와 같이 한 페이지에서만 사용하는 상수를 넣기엔 좀 그럴 것 같습니다.
페이지 폴더에 constants 파일을 만들어서 페이지 내에서 사용하는 상수는 모두 constants 파일에 넣는 것이 좋을지 아니면 한 컴포넌트에서만 사용되는 상수라면 컴포넌트 파일에 두면 좋을지는 논의해보면 좋을 것 같아요!

jaypedia replied:
확실히 응집도를 생각한다면, 별도의 파일 내부에 놓는 것은 별로 좋아 보이지 않겠네요! 
constants 파일을 만드는 것도 나쁘지 않을 것 같아요. constants가 길어지면, 컴포넌트 파일 길이가 너무 길어져서요.

soralee2821 replied:
확실히 상수가 많아지면, 컴포넌트 파일 길이도 길어지긴 하죠;
페이지 폴더 내부에 constants 파일을 만들어서 상수를 관리하는 방법 좋습니다!
그렇게 수정해서 커밋할께요~

jaypedia reviewed (changes requested):
회원 탈퇴 기능 구현하느라 고생 많으셨어요! 
타이머도 잘 구현하시고, 정리도 깔끔하게 잘 하셔서 이번 PR도 파악하기가 너무 좋았습니다 👍
몇 가지 제안 사항에 대해서 논의를 하고 반영 후 merge 진행하면 될거같아요~

soralee2821 committed:
[#61] Refactor: correct the content of TEXT_CONTENTS for grammer

soralee2821 committed:
[#61] Refactor: move constants to constant file in DeleteAccount folder

soralee2821 committed:
[#61] Refactor: change img tag to svg tag

- 이미지 로딩 속도가 느리므로 img 태그를 svg 태그로 변경
- 아이콘이 메세지 내용과 같은 줄에 있게 하기 위한 CSS를 변경된 svg 태그에도 그대로 적용하기 위해서, svg 태그에 id를 부여하고 index.css 파일에서 CSS를 적용함

jaypedia commented on src/index.css:
```diff
@@ -28,3 +28,8 @@
   border-top: 1px solid #555;
 }
+
+#success-icon {
```
icon 자체에 style을 부여할 수 있는데, 그렇게 하지 않으시고 이런 방식으로 스타일을 주신 이유가 따로 있으신가요?
global한 index.css 파일에 success-icon만을 위한 스타일을 놓는다는 게 살짝 위치상으로 어색해 보이기도 해요!

soralee2821 replied:
className 속성이 부여 안되서 CSS 파일로 따로 설정해야 하는 줄 알았습니다; HTML 태그여서 style 속성을 부여할 수 있는 것을 미처 생각을 못했네요. 
저도 index.css에 작성하는 것이 맘에 걸렸는데 더 좋은 방법 알려주셔서 감사합니다!

jaypedia commented on src/pages/DeleteAccount/constants.ts:
```diff
@@ -0,0 +1,26 @@
+export const TEXT_CONTENTS = {
+  invalidAuthenticationCodeDescription: '유효하지 않은 인증번호입니다.',
+  modal: {
+    main: '회원탈퇴',
```
`src/pages/DeleteAccount/index.tsx`여기서는 '회원 탈퇴'라고 띄어쓰기가 되어 있는데 여기는 붙여있네요! 일관성 + 문법에 맞게 띄어쓰기 적용하면 좋겠습니다~

soralee2821 replied:
와이어프레임의 'CJ ONE 회원 탈퇴 모달 예시'에서 회원 탈퇴 문구가 붙어 있어서 따라서 작성했는데 회원 탈퇴 페이지의 문구와 일관성이 없었네요;
'회원 탈퇴'로 모두 띄어쓰기를 적용해서 수정했습니다.

soralee2821 committed:
[#61] Refactor: move CSS properties from index.css file to style tag in SuccessIcon

soralee2821 committed:
[#61] Refactor: correct the contents of TEXT_CONTENTS for grammer

- '회원탈퇴'를 맞춤법에 맞게 '회원 탈퇴'로 변경

jaypedia commented on src/pages/DeleteAccount/index.tsx:
```diff
@@ -0,0 +1,59 @@
+        content: '이메일로 인증 코드가 발송되었습니다.',
+        duration: 2,
+        style: {
+          marginTop: '75vh',
```
`marginTop: 75vh`로 했을 때 전체 화면을 기준으로 하기 때문에 문제가 발생한다고 하신 것 같은데 잘 해결 되셨나요~?

soralee2821 replied:
Toast의 marginTop은 전체화면 기준으로 배치되기 때문에 확대/축소가 67% 미만일 경우 toast가 Root 컴포넌트 밖으로 벗어났습니다. 그런데 사용자가 그렇게 축소해서 볼 경우가 적다고 생각해서 그대로 두었습니다.
